[Lian Guo Yang] iP

LianGuoYang opened the pull request:
## Duke Pro 

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/LianGuoYang/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉 

And it is **FREE**!

Features:

- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)  

---

If you Java programmer, you can use it to practice Java too. Here's the <span style="color: red">`main`</span>  method:

```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

LianGuoYang committed:
Implemented Level-1. ADD added Greet, Echo and Exit feature for Duke bot

LianGuoYang committed:
Implemented Level-2. ADD Added Add and List feature to add tasks and print all current tasks

LianGuoYang committed:
Implemented Level-3. Added parent class Command and subclasses command for different commands. Added Echo class to analyse input string. Added class Tasks.

LianGuoYang committed:
Rename Tasks to Task

LianGuoYang committed:
Implemented Level-4. Added Task subclasses Todo, Event, Deadline.

LianGuoYang committed:
Changed minor output string

LianGuoYang committed:
A-TextUiTesting: Automated Text UI Testing

LianGuoYang committed:
Implemented Level-5. Added exceptions to deal with errors for AddCmd

LianGuoYang committed:
Implemented Level-6. Added delete feature

LianGuoYang committed:
A-Enums. Add enums for tasks

LianGuoYang committed:
Implemented Level-7. Add Save feature

LianGuoYang committed:
Implemented Level-8. Add datetime to deadline and event tasks

LianGuoYang committed:
Merge branch 'branch-Level-7' and 'branch-Level-8' to master branch

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Event.java

LianGuoYang committed:
Implemented A-MoreOOP
Add Ui, Storage, Parser, TaskList and Command subclasses

LianGuoYang committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

LianGuoYang committed:
Add checkstyle

LianGuoYang committed:
Implement A-Packages. All classes in duke package, and divided
into multiple packages, duke.command, duke.exception and duke.task.

LianGuoYang committed:
MarkCommand class : Add getNum method
UnmarkCommand class : Add getNum method
Parser class : Throw exceptions when index is empty

LianGuoYang committed:
Implement A-JUnit
Add JUnit tests for Parser, TaskList, AddCommand, Todo

LianGuoYang committed:
Implement Level 9
FindCommand to find tasks that have the same name as the keyword

LianGuoYang committed:
Add Javadocs for Commands

LianGuoYang committed:
Add Javadocs for DukeException and Task

LianGuoYang committed:
Add Javadocs for all classes

LianGuoYang committed:
Implement A-CodingStandard for all classes

LianGuoYang committed:
Merge branch 'branch-Level-9'

LianGuoYang committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/Ui.java

LianGuoYang committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/Ui.java
#	src/main/java/duke/command/AddCommand.java
#	src/main/java/duke/command/Command.java
#	src/main/java/duke/command/DeleteCommand.java
#	src/main/java/duke/command/ExitCommand.java
#	src/main/java/duke/command/ListCommand.java
#	src/main/java/duke/command/MarkCommand.java
#	src/main/java/duke/command/UnmarkCommand.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/task/Task.java

DanielLimWeiEn commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,59 @@
+
+import java.io.IOException;
+
+import duke.command.Command;
```
I really like how you have kept the ordering of import statements very consistent in all files and that the imports were all explicit 👍. It made reading and following the code extremely smooth.

DanielLimWeiEn commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,65 @@
+     * Add a new task into the TaskList
+     * @param newTask new task added
+     */
+    public void add(Task newTask) {
```
Do you think it would be better to name this addTask and deleteTask? Perhaps to give the method a more descriptive name? 🤔

DanielLimWeiEn commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,176 @@
+     *
+     * @param message Error message
+     */
+    public void showLoadingError(String message) {
```
Would it perhaps be a better idea to pass in the Exception object as an argument to this method, maybe to make the method more descriptive?

DanielLimWeiEn commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    /**
+     * Execute the delete command and remove the task at index.
```
I like that you added descriptive JavaDoc comments for all of the public methods 👍. They are descriptive and make reading the code easy since it is clear what the methods are supposed to do.

DanielLimWeiEn reviewed (commented):
Overall looks very good 👍. The code is clear and easy to follow, with descriptive names and sufficient documentation to explain what is being done. And I like that the code has been logically separated into different parts. Just a little bit of nitpicking by me 😅 is all.

LianGuoYang committed:
Implement Level-10: GUI

LianGuoYang committed:
Implement A-CheckStyle

LianGuoYang committed:
Rename method name
add and remove to addTask and removeTask

LianGuoYang committed:
Merge branch 'branch-Level-10'

debwy commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,168 @@
+                }
+
+                String[] info = line.split(" / ", 4);
+                switch (info[0]) {
+                case "T":
+                    tasks.add(new Todo(TaskType.TODO, info[2], info[1].equals("1")));
+                    break;
+                case "D":
+                    tasks.add(new Deadline(TaskType.DEADLINE, info[2], info[1].equals("1"), info[3]));
+                    break;
+                case "E":
+                    tasks.add(new Event(TaskType.EVENT, info[2], info[1].equals("1"), info[3]));
+                    break;
+                default:
+                    throw new DukeException("Invalid task");
+                }
```
I like how even for switch cases without many cases you still used it instead of if-else statements, which makes your code more readable!

debwy commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    /**
+     * Mark the task as done
```
(This is just a nitpick but) you might want to rephrase the javadoc header comments in the style "marks", "unmarks", etc. to comply with the coding standard document

LianGuoYang replied:
Yea thanks for reminding and spotted the mistakes in my code!

debwy reviewed (commented):
Generally the code is clean and easy to read, the javadoc statements are descriptive and properly included, and there are no notable code quality violations (both naming and logic flow made sense) -- overall nice job 👍

LianGuoYang committed:
Add minor details to GUI

LianGuoYang committed:
Edit Javadocs according to suggestions from debwy and Daniels

LianGuoYang committed:
Implement A-Varags

LianGuoYang committed:
Merge branch 'branch-A-Varargs'

LianGuoYang committed:
Implement A-Assertions:

Assertions: define assumptions

There are multiple assumptions made in many methods.

Assertions help document important assumptions that should
be hold at various points in code.

Add assert statements to prevent errors made.

Let's add assertions to the classes:
* duke.Duke
* duke.Parser
* duke.Storage
* duke.TaskList
* duke.Ui
* all of the commands in duke.command package

LianGuoYang committed:
Implement A-CodeQuality

CodeQuality: to make the code readable and let others to understand
my code easier.

There is a long method and deep nesting in AddCommand which makes
it difficult to understand. There are multiple magic literals with
no meaning.

Implement SLAP to remove long method and deep nesting in AddCommand.
Change magic literals to meaningful variable names.

SLAP and meaningful variable names allow others to understand the
code better.

Let's
* Follow SLAP to remove long methods and deep nesting.
* remove magic literals and change them to meaningful names.

LianGuoYang committed:
Merge pull request #1 from LianGuoYang/branch-A-Assertions

Implement A-Assertions

LianGuoYang committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/Ui.java
#	src/main/java/duke/command/FindCommand.java

LianGuoYang committed:
Merge pull request #3 from LianGuoYang/branch-A-CodeQuality

Implement A-CodeQuality

LianGuoYang committed:
Implement B-Snooze for BCD-Extension

It is very inconvenient to delete a task and add a task with
same details but a different date.

Snooze command to change task to a later date.

Allow users to snooze deadline and event task to a later date.

Let's, implement a snooze command to allow users to change
their task to a later date.

LianGuoYang committed:
Merge pull request #5 from LianGuoYang/branch-BCD-Extension

Implement B-Snooze for BCD-Extension

LianGuoYang committed:
Merge branch 'master' into branch-A-CodeQuality

LianGuoYang committed:
Fix code quality according to feedback

LianGuoYang committed:
Merge pull request #7 from LianGuoYang/branch-A-CodeQuality

Fix code quality according to feedback

LianGuoYang committed:
Implement A-MoreErrorHandling

Add more exceptions to catch errors.

LianGuoYang committed:
Merge branch 'branch-A-MoreErrorHandling'

LianGuoYang committed:
Improve GUI

Delay exit program to show full exit message.
Show exceptions messages.

LianGuoYang committed:
Improve text saving for each command

LianGuoYang committed:
Merge branch 'master' into branch-A-BetterGui

LianGuoYang committed:
Implement A-BetterGui

Add different background to user and response.

LianGuoYang committed:
Merge branch 'branch-A-BetterGui'

LianGuoYang committed:
Implement A-Personality

LianGuoYang committed:
Fix minor bugs

LianGuoYang committed:
Upload Ui png

LianGuoYang committed:
Implement A-UserGuide and write user guide

LianGuoYang committed:
Fix wrong path for Launcher

LianGuoYang committed:
Fix user guide
